File: web/src/api/frpServer.ts
import request from './request';

export interface FrpsMetrics {
  client_counts: number;
  proxy_counts: Record<string, number>;
  total_proxies: number;
  traffic_in: number;
  traffic_out: number;
  cpu_seconds: number;
  memory_bytes: number;
  start_time: number;
  uptime: number;
  goroutines: number;
}

export interface ServerMetricsHistory {
  id: number;
  server_id: number;
  cpu_percent: number;
  memory_bytes: number;
  traffic_in: number;
  traffic_out: number;
  record_time: string;
}

export interface FrpServer {
  id?: number;
  name: string;
  server_type?: 'local' | 'remote';
  host: string;
  dashboard_port: number;
  dashboard_user: string;
  dashboard_pwd: string;
  bind_port: number;
  token?: string;
  ssh_host?: string;
  ssh_port?: number;
  ssh_user?: string;
  ssh_password?: string;
  install_path?: string;
  mirror_id?: number;
  enabled: boolean;
  status?: string;
  pid?: number;
  version?: string;
  binary_path?: string;
  config_path?: string;
  last_sync_time?: string;
  last_error?: string;
  created_at?: string;
  updated_at?: string;
}

export const frpServerApi = {
  getAll: () => request.get<FrpServer[]>('/frp-servers'),
  
  getById: (id: number) => request.get<FrpServer>(`/frp-servers/${id}`),
  
  create: (data: FrpServer) => request.post<FrpServer>('/frp-servers', data),
  
  update: (id: number, data: FrpServer) => request.put<FrpServer>(`/frp-servers/${id}`, data),
  
  delete: (id: number, removeInstallation?: boolean) => request.delete(`/frp-servers/${id}`, { params: { remove_installation: removeInstallation } }),
  
  testConnection: (data: FrpServer) => request.post('/frp-servers/test', data),
  
  start: (id: number) => request.post(`/frp-servers/${id}/start`),
  
  stop: (id: number) => request.post(`/frp-servers/${id}/stop`),
  
  restart: (id: number) => request.post(`/frp-servers/${id}/restart`),
  
  getStatus: (id: number) => request.get<{ status: string }>(`/frp-servers/${id}/status`),
  
  download: (id: number, version?: string) => request.post(`/frp-servers/${id}/download`, { version }),
  
  testSSH: (id: number) => request.post(`/frp-servers/${id}/test-ssh`),
  
  remoteInstall: (id: number, mirrorId?: number) => request.post(`/frp-servers/${id}/remote-install`, { mirror_id: mirrorId }),
  
  remoteStart: (id: number) => request.post(`/frp-servers/${id}/remote-start`),
  
  remoteStop: (id: number) => request.post(`/frp-servers/${id}/remote-stop`),
  
  remoteRestart: (id: number) => request.post(`/frp-servers/${id}/remote-restart`),
  
  remoteUninstall: (id: number) => request.post(`/frp-servers/${id}/remote-uninstall`),
  
  remoteGetLogs: (id: number, lines?: number) => request.get<{ logs: string }>(`/frp-servers/${id}/remote-logs`, { params: { lines } }),
  
  remoteGetVersion: (id: number) => request.get<{ version: string }>(`/frp-servers/${id}/remote-version`),
  
  getLocalVersion: (id: number) => request.get<{ version: string }>(`/frp-servers/${id}/local-version`),
  
  remoteReinstall: (id: number, regenerateAuth?: boolean, mirrorId?: number) => request.post(`/frp-servers/${id}/remote-reinstall`, { regenerate_auth: regenerateAuth, mirror_id: mirrorId }),
  
  remoteUpgrade: (id: number, version?: string, mirrorId?: number) => request.post(`/frp-servers/${id}/remote-upgrade`, { version, mirror_id: mirrorId }),
  
  getRunningTask: (id: number) => request.get<{ running: boolean; operation?: string }>(`/frp-servers/${id}/running-task`),
  
  parseConfig: (config: string) => request.post<{ bind_port: number; token: string; host: string; dashboard_port: number; dashboard_user: string; dashboard_pwd: string }>('/frp-servers/parse-config', { config }),
  
  getMetrics: (id: number) => request.get<FrpsMetrics>(`/frp-servers/${id}/metrics`),
  
  getMetricsHistory: (id: number, days?: number) => request.get<ServerMetricsHistory[]>(`/frp-servers/${id}/metrics-history`, { params: { days } }),
};

File: backend/internal/model/frp_server.go
/*
 * @Author              : 寂情啊
 * @Date                : 2025-11-19 17:06:28
 * @LastEditors         : 寂情啊
 * @LastEditTime        : 2025-11-21 14:04:04
 * @FilePath            : frp-web-testbackendinternalmodelfrp_server.go
 * @Description         : 说明
 * 倾尽绿蚁花尽开，问潭底剑仙安在哉
 */
package model

import "time"

type FrpServerStatus string

const (
	StatusStopped  FrpServerStatus = "stopped"
	StatusStarting FrpServerStatus = "starting"
	StatusRunning  FrpServerStatus = "running"
	StatusStopping FrpServerStatus = "stopping"
	StatusError    FrpServerStatus = "error"
)

type ServerType string

const (
	ServerTypeLocal  ServerType = "local"
	ServerTypeRemote ServerType = "remote"
)

type FrpServer struct {
	ID            uint            `json:"id" gorm:"primaryKey"`
	Name          string          `json:"name" gorm:"uniqueIndex;size:100;not null"`
	ServerType    ServerType      `json:"server_type" gorm:"size:20;default:'local'"`
	Host          string          `json:"host" gorm:"size:255;not null"`
	DashboardPort int             `json:"dashboard_port" gorm:"not null;default:7500"`
	DashboardUser string          `json:"dashboard_user" gorm:"size:100"`
	DashboardPwd  string          `json:"dashboard_pwd" gorm:"size:255"`
	BindPort      int             `json:"bind_port" gorm:"default:7000"`
	Token         string          `json:"token" gorm:"size:64"`
	SSHHost       string          `json:"ssh_host" gorm:"size:255"`
	SSHPort       int             `json:"ssh_port" gorm:"default:22"`
	SSHUser       string          `json:"ssh_user" gorm:"size:100"`
	SSHPassword   string          `json:"ssh_password" gorm:"size:500"`
	InstallPath   string          `json:"install_path" gorm:"size:500;default:'/opt/frps'"`
	MirrorID      *uint           `json:"mirror_id"`
	Enabled       bool            `json:"enabled" gorm:"default:true"`
	Status        FrpServerStatus `json:"status" gorm:"size:20;default:'stopped'"`
	PID           int             `json:"pid" gorm:"default:0"`
	Version       string          `json:"version" gorm:"size:50"`
	BinaryPath    string          `json:"binary_path" gorm:"size:500"`
	ConfigPath    string          `json:"config_path" gorm:"size:500"`
	LastSyncTime  *time.Time      `json:"last_sync_time"`
	LastError     string          `json:"last_error" gorm:"type:text"`
	CreatedAt     time.Time       `json:"created_at"`
	UpdatedAt     time.Time       `json:"updated_at"`
}
